File: ReactNative/PerimeterX_SDK_Demo/ios/PerimeterX_SDK_Demo/PerimeterXModule.m
#import "PerimeterXModule.h"
@import PerimeterX_SDK;

static NSString *pxNewHeaders = @"PxNewHeaders";
static NSString *pxChallengeResult = @"PxChallengeResult";
static NSString *pxSolved = @"solved";
static NSString *pxCancelled = @"cancelled";
static NSString *pxFalse = @"false";

@implementation PerimeterXModule

// MARK: - NSObject

- (instancetype)init {
  self = [super init];
  shared = self;
  return self;
}

// MARK: - RCTEventEmitter

- (NSArray<NSString *> *)supportedEvents {
  return @[pxNewHeaders, pxChallengeResult];
}

// MARK: - PX Module

static PerimeterXModule *shared = nil;

+ (PerimeterXModule *)shared {
  return shared;
}

- (void)handleUpdatedHeaders:(NSDictionary<NSString *,NSString *> *)headers {
  NSData *data = [NSJSONSerialization dataWithJSONObject:headers options:0 error:nil];
  NSString *json = [[NSString alloc] initWithData:data encoding:NSUTF8StringEncoding];
  [self sendEventWithName:pxNewHeaders body:json];
}

- (void)handleChallengeSolvedEvent {
  [self sendEventWithName:pxChallengeResult body:pxSolved];
}

- (void)handleChallengeCancelledEvent {
  [self sendEventWithName:pxChallengeResult body:pxCancelled];
}

RCT_EXPORT_MODULE(PerimeterXModule);

RCT_REMAP_METHOD(getHTTPHeaders,
                 resolver:(RCTPromiseResolveBlock)resolve
                 rejecter:(RCTPromiseRejectBlock)reject) {
  NSDictionary<NSString *, NSString *> *headers = [PerimeterX headersForURLRequestForAppId:nil];
  NSData *data = [NSJSONSerialization dataWithJSONObject:headers options:0 error:nil];
  NSString *json = [[NSString alloc] initWithData:data encoding:NSUTF8StringEncoding];
  resolve(@[json]);
}

RCT_REMAP_METHOD(handleResponse,
                 rsponse:(NSString *)response
                 code:(NSInteger)code
                 url:(NSString *)url
                 resolver:(RCTPromiseResolveBlock)resolve
                 rejecter:(RCTPromiseRejectBlock)reject) {
  NSData *data = [response dataUsingEncoding:NSUTF8StringEncoding];
  NSHTTPURLResponse *httpURLResponse = [[NSHTTPURLResponse alloc] initWithURL:[NSURL URLWithString:url] statusCode:code HTTPVersion:nil headerFields:nil];
  BOOL handled = [PerimeterX handleResponseWithResponse:httpURLResponse data:data forAppId:nil callback:^(enum PerimeterXChallengeResult result) {
    resolve((result == PerimeterXChallengeResultSolved ? pxSolved : pxCancelled));
  }];
  if (!handled) {
    resolve(pxFalse);
  }
}

@end

File: ReactNative/HUMAN_Demo/android/app/src/main/java/com/perimeterx_sdk_demo/PerimeterXModule.java
package com.perimeterx_sdk_demo;

import androidx.annotation.NonNull;

import com.facebook.react.bridge.Promise;
import com.facebook.react.bridge.ReactApplicationContext;
import com.facebook.react.bridge.ReactContextBaseJavaModule;
import com.facebook.react.bridge.ReactMethod;
import com.facebook.react.modules.core.DeviceEventManagerModule;
import com.perimeterx.mobile_sdk.PerimeterX;
import com.perimeterx.mobile_sdk.PerimeterXChallengeResult;

import org.json.JSONObject;

import java.util.HashMap;

public class PerimeterXModule extends ReactContextBaseJavaModule {

    static PerimeterXModule shared = null;

    String pxNewHeaders = "PxNewHeaders";
    String pxChallengeResult = "PxChallengeResult";
    String pxSolved = "solved";
    String pxCancelled = "cancelled";
    String pxFalse = "false";

    // PX Module

    PerimeterXModule(ReactApplicationContext context) {
        super(context);
    }

    @NonNull
    @Override
    public String getName() {
        return "PerimeterXModule";
    }

    public void handleUpdatedHeaders(HashMap<String, String> headers) {
        if (!this.getReactApplicationContext().hasCatalystInstance()) {
            return;
        }
        JSONObject json = new JSONObject(headers);
        this.getReactApplicationContext()
                .getJSModule(DeviceEventManagerModule.RCTDeviceEventEmitter.class)
                .emit(pxNewHeaders, json.toString());
    }

    public void handleChallengeSolvedEvent() {
        if (!this.getReactApplicationContext().hasCatalystInstance()) {
            return;
        }
        this.getReactApplicationContext()
                .getJSModule(DeviceEventManagerModule.RCTDeviceEventEmitter.class)
                .emit(pxChallengeResult, pxSolved);
    }

    public void handleChallengeCancelledEvent() {
        if (!this.getReactApplicationContext().hasCatalystInstance()) {
            return;
        }
        this.getReactApplicationContext()
                .getJSModule(DeviceEventManagerModule.RCTDeviceEventEmitter.class)
                .emit(pxChallengeResult, pxCancelled);
    }

    @ReactMethod
    public void getHTTPHeaders(Promise promise) {
        JSONObject json = new JSONObject(PerimeterX.INSTANCE.headersForURLRequest(null));
        promise.resolve(json.toString());
    }

    @ReactMethod
    public void handleResponse(String response, Integer code, String url, Promise promise) {
        boolean handled = PerimeterX.INSTANCE.handleResponse(response, null, result -> {
            promise.resolve(result == PerimeterXChallengeResult.SOLVED ? pxSolved : pxCancelled);
            return null;
        });
        if (!handled) {
            promise.resolve(pxFalse);
        }
    }
}
